x/session/keeper/session_hydrator.go

```diff
@@ -0,0 +1,207 @@
+
+	if err := k.hydrateSessionID(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate the session ID: %v", err)
+	}
+	logger.Info("Finished hydrating session ID: %s", sh.sessionHeader.SessionId)
+
+	if err := k.hydrateSessionApplication(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate application for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session application: %+v", sh.session.Application)
+
+	if err := k.hydrateSessionSuppliers(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate suppliers for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session suppliers: %+v")
+
+	sh.session.Header = sh.sessionHeader
+	sh.session.SessionId = sh.sessionHeader.SessionId
+
+	return sh.session, nil
+}
+
+// hydrateSessionMetadata hydrates metadata related to the session such as the height at which the session started, its number, the number of blocks per session, etc..
+func (k Keeper) hydrateSessionMetadata(ctx sdk.Context, sh *sessionHydrator) error {
+	sh.session.NumBlocksPerSession = NumBlocksPerSession
+	sh.session.SessionNumber = int64(sh.blockHeight/NumBlocksPerSession) + 1
+	sh.sessionHeader.SessionStartBlockHeight = sh.blockHeight - (sh.blockHeight % NumBlocksPerSession)
+	return nil
+}
+
+// hydrateSessionID use both session and on-chain data to determine a unique session ID
+func (k Keeper) hydrateSessionID(ctx sdk.Context, sh *sessionHydrator) error {
+	// TODO_TECHDEBT: Need to retrieve the block hash at SessionStartBlockHeight, NOT THE CURRENT ONE
+	prevHashBz := ctx.HeaderHash()
+	appPubKeyBz := []byte(sh.sessionHeader.ApplicationAddress)
+	serviceIdBz := []byte(sh.sessionHeader.ServiceId.Id)
+	sessionHeightBz := make([]byte, 8)
+	binary.LittleEndian.PutUint64(sessionHeightBz, uint64(sh.sessionHeader.SessionStartBlockHeight))
+
+	sh.sessionIdBz = concatWithDelimiter(SessionIDComponentDelimiter, prevHashBz, serviceIdBz, appPubKeyBz, sessionHeightBz)
+	sh.sessionHeader.SessionId = hex.EncodeToString(sha3Hash(sh.sessionIdBz))
+
+	return nil
+}
+
+// hydrateSessionApplication hydrates the full Application actor based on the address provided
+func (k Keeper) hydrateSessionApplication(ctx sdk.Context, sh *sessionHydrator) error {
+	app, appIsFound := k.appKeeper.GetApplication(ctx, sh.sessionHeader.ApplicationAddress)
+	if !appIsFound {
+		return sdkerrors.Wrapf(types.ErrHydratingSession, "failed to find session application")
+	}
+	sh.session.Application = &app
+	return nil
+}
+
+// hydrateSessionSuppliers finds the suppliers that are staked at the session height and populates the session with them
+func (k Keeper) hydrateSessionSuppliers(ctx sdk.Context, sh *sessionHydrator) error {
+	logger := k.Logger(ctx).With("method", "hydrateSessionSuppliers")
+
+	// TODO_TECHDEBT(@Olshansk): Need to retrieve the suppliers at SessionStartBlockHeight,
```

Comment:
@bryanchriswhite Would the BlockClient + ReplayObservable (where N = SessionLength) enable this?

Reply:
Resolving and just updating a TODO.